File: src/medium/valid_sudoku.rs
struct Solution;

impl Solution {
    // https://leetcode.com/problems/valid-sudoku
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows = [0u16; 9];
        let mut cols = [0u16; 9];
        let mut boxes = [0u16; 9];

        for r in 0..9 {
            for (c, &ch) in board[r].iter().enumerate() {
                if ch == '.' {
                    continue;
                }
                if !('1'..='9').contains(&ch) {
                    return false;
                }

                let bit = 1u16 << ((ch as u8 - b'1') as u16);
                let b = (r / 3) * 3 + (c / 3);

                if (rows[r] & bit) != 0 || (cols[c] & bit) != 0 || (boxes[b] & bit) != 0 {
                    return false;
                }

                rows[r] |= bit;
                cols[c] |= bit;
                boxes[b] |= bit;
            }
        }
        true
    }
}

#[test]
fn test() {
    let test_cases = [
        (
            vec![
                vec!['5', '3', '.', '.', '7', '.', '.', '.', '.'],
                vec!['6', '.', '.', '1', '9', '5', '.', '.', '.'],
                vec!['.', '9', '8', '.', '.', '.', '.', '6', '.'],
                vec!['8', '.', '.', '.', '6', '.', '.', '.', '3'],
                vec!['4', '.', '.', '8', '.', '3', '.', '.', '1'],
                vec!['7', '.', '.', '.', '2', '.', '.', '.', '6'],
                vec!['.', '6', '.', '.', '.', '.', '2', '8', '.'],
                vec!['.', '.', '.', '4', '1', '9', '.', '.', '5'],
                vec!['.', '.', '.', '.', '8', '.', '.', '7', '9'],
            ],
            true,
        ),
        (
            vec![
                vec!['8', '3', '.', '.', '7', '.', '.', '.', '.'],
                vec!['6', '.', '.', '1', '9', '5', '.', '.', '.'],
                vec!['.', '9', '8', '.', '.', '.', '.', '6', '.'],
                vec!['8', '.', '.', '.', '6', '.', '.', '.', '3'],
                vec!['4', '.', '.', '8', '.', '3', '.', '.', '1'],
                vec!['7', '.', '.', '.', '2', '.', '.', '.', '6'],
                vec!['.', '6', '.', '.', '.', '.', '2', '8', '.'],
                vec!['.', '.', '.', '4', '1', '9', '.', '.', '5'],
                vec!['.', '.', '.', '.', '8', '.', '.', '7', '9'],
            ],
            false,
        ),
    ];

    for (input, expected) in test_cases {
        assert_eq!(
            Solution::is_valid_sudoku(input.clone()),
            expected,
            "failed for input {input:?}",
        );
    }
}

File: src/medium/valid-sudoku.go
package medium

import "slices"

// https://leetcode.com/problems/valid-sudoku
func isValidSudoku(board [][]byte) bool {
	var rows, cols, blocks [9]uint16

	for i := range 9 {
		for j := range 9 {
			elem := board[i][j]
			if elem == '.' {
				continue
			}
			if elem < '1' || elem > '9' {
				return false
			}

			bit := uint16(1) << (elem - '1')

			blockIdx := (i/3)*3 + (j / 3)

			if rows[i]&bit != 0 || cols[j]&bit != 0 || blocks[blockIdx]&bit != 0 {
				return false
			}

			rows[i] |= bit
			cols[j] |= bit
			blocks[blockIdx] |= bit
		}
	}

	return true
}

func isValidSudokuOld(board [][]byte) bool {
	set := make(map[byte]struct{}, 9)

	check := func(elem byte) bool {
		if elem == '.' {
			return false
		}
		if elem < '1' || elem > '9' {
			return true
		}
		if _, ok := set[elem]; ok {
			return true
		}
		set[elem] = struct{}{}
		return false
	}

	for _, row := range board {
		if slices.ContainsFunc(row, check) {
			return false
		}
		clear(set)
	}

	for j := range 9 {
		for i := range 9 {
			if check(board[i][j]) {
				return false
			}
		}
		clear(set)
	}

	for x := 1; x <= 3; x++ {
		for y := 1; y <= 3; y++ {
			for j := (x - 1) * 3; j < x*3; j++ {
				for i := (y - 1) * 3; i < y*3; i++ {
					if check(board[i][j]) {
						return false
					}
				}
			}
			clear(set)
		}
	}

	return true
}
